src/openvic-simulation/modifier/ModifierManager.cpp

```diff
@@ -686,181 +661,140 @@ bool ModifierManager::parse_scripts(DefinitionManager const& definition_manager)
-	};
-}
+	const std::string flat_identifier = get_flat_identifier(prefix, key);
+	ModifierEffect const* effect = technology_modifier_effects.get_item_by_identifier(flat_identifier);
+	if (effect != nullptr) {
+		return _add_modifier_cb(modifier_value, effect, value);
+	} else {
+		Logger::error("Could not find flattened modifier: ", flat_identifier);
+		return false;
+	}
+};
 
-node_callback_t ModifierManager::expect_validated_modifier_value_and_default(
-	callback_t<ModifierValue&&> modifier_callback, Modifier::modifier_type_t type, key_value_callback_t default_callback,
-	ModifierEffectValidator auto effect_validator
+bool ModifierManager::_add_modifier_cb(
+	ModifierValue& modifier_value,
+	ModifierEffect const* const effect,
+	const ast::NodeCPtr value
 ) const {
-	return [this, modifier_callback, type, default_callback, effect_validator](ast::NodeCPtr root) -> bool {
-		ModifierValue modifier;
-
-		bool ret = expect_dictionary_reserve_length(
-			modifier.values,
-			_modifier_effect_callback(modifier, type, default_callback, effect_validator)
-		)(root);
-
-		ret &= modifier_callback(std::move(modifier));
-
-		return ret;
-	};
+	if (effect->has_no_effect()) {
+		Logger::warning("This modifier does nothing: ", effect->get_identifier());
+	}
+	return expect_fixed_point(map_callback(modifier_value.values, effect))(value);
 }
 
-node_callback_t ModifierManager::expect_validated_modifier_value(
-	callback_t<ModifierValue&&> modifier_callback, Modifier::modifier_type_t type,
-	ModifierEffectValidator auto effect_validator
+NodeTools::key_value_callback_t ModifierManager::_expect_modifier_effect(
+	modifier_effect_registry_t const& registry,
+	ModifierValue& modifier_value
 ) const {
-	return expect_validated_modifier_value_and_default(modifier_callback, type, key_value_invalid_callback, effect_validator);
+	return _expect_modifier_effect_with_fallback(registry, modifier_value, key_value_invalid_callback);
 }
-
-node_callback_t ModifierManager::expect_modifier_value_and_default(
-	callback_t<ModifierValue&&> modifier_callback, Modifier::modifier_type_t type, key_value_callback_t default_callback
+NodeTools::key_value_callback_t ModifierManager::_expect_modifier_effect_with_fallback(
+	modifier_effect_registry_t const& registry,
+	ModifierValue& modifier_value,
+	const NodeTools::key_value_callback_t fallback
 ) const {
-	return expect_validated_modifier_value_and_default(
-		modifier_callback, type, default_callback, [](ModifierEffect const&) -> bool {
-			return true;
+	return [this, &registry, &modifier_value, fallback]
+		(const std::string_view key, const ast::NodeCPtr value) {
+		if (dryad::node_has_kind<ast::ListValue>(value) && complex_modifiers.contains(key)) {
+			return expect_dictionary([this, &modifier_value, &key]
+				(const std::string_view inner_key, const ast::NodeCPtr inner_value) -> bool{
+				return _add_flattened_modifier_cb(modifier_value, key, inner_key, inner_value);
+			})(value);
 		}
```

Comment:
```suggestion
	return [this, &registry, &modifier_value, fallback](const std::string_view key, const ast::NodeCPtr value) -> bool {
		if (dryad::node_has_kind<ast::ListValue>(value) && complex_modifiers.contains(key)) {
			return expect_dictionary([this, &modifier_value, key](
				const std::string_view inner_key, const ast::NodeCPtr inner_value
			) -> bool {
				return _add_flattened_modifier_cb(modifier_value, key, inner_key, inner_value);
			})(value);
		}
```
Formatting + capture `key` by value rather than reference + return type of outer lambda function